{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Imports\n",
    "All imports needed here"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import sys\n",
    "import subprocess\n",
    "import re\n",
    "import pickle\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "from sklearn.model_selection import train_test_split, KFold"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Xstr function\n",
    "Function that returns an empty string if the object \"s\" is None, or the str value of that object\n",
    "\n",
    "Parameters:\n",
    "\n",
    "&emsp;__s__: Object which you want to convert to str<br>\n",
    "\n",
    "Return value:\n",
    "\n",
    "&emsp;'' if s is None or a string with the value of s in other case."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "def xstr(s):\n",
    "    if s is None:\n",
    "        return ''\n",
    "    return str(s)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## GetHandlerToGenotypeData function\n",
    "Function that returns an empty string if the object \"s\" is None, or the str value of that object\n",
    "\n",
    "Parameters:\n",
    "\n",
    "&emsp;__geno__: Full path to the plink files, not including the extension.<br>\n",
    "&emsp;__covs__: A covariates file with the minimal information about samples<br>\n",
    "&emsp;__predictor__: What you want to predict<br>\n",
    "&emsp;__id__: Name of the column that represents the id<br>\n",
    "&emsp;__fid__: Name of the column that represents the fid<br>\n",
    "&emsp;__pheno__: Name of the pheno file<br>\n",
    "&emsp;__verify__: If you want to ensure the files exist (By default True).<br>\n",
    "\n",
    "Return value:\n",
    "\n",
    "&emsp;A handler with the necessary data (geno, pheno, covs)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "def getHandlerToGenotypeData(geno,\n",
    "                             covs,\n",
    "                             predictor,\n",
    "                             id,\n",
    "                             fid,\n",
    "                             pheno,\n",
    "                             verify=True):\n",
    "\n",
    "    # Ensure the geno files exist\n",
    "    if verify and\\\n",
    "        (not os.path.isfile(geno + \".bed\") or\n",
    "         not os.path.isfile(geno + \".bim\") or\n",
    "         not os.path.isfile(geno + \".fam\")):\n",
    "\n",
    "        sys.exit(\"Files not found: \" + geno)\n",
    "\n",
    "    # Ensure the covs file exists\n",
    "    if not os.path.isfile(covs + \".cov\"):\n",
    "        sys.exit(\"Files not found: \" + covs)\n",
    "\n",
    "    cdata = pd.read_csv(covs + \".cov\", header=0, delim_whitespace=True)\n",
    "\n",
    "    phenofile = pheno + \".pheno\"\n",
    "\n",
    "    # Ensure pheno file has the predictor, id and fid\n",
    "    if not os.path.isfile(phenofile):\n",
    "        if predictor not in cdata.columns:\n",
    "            sys.exit(\"Error: \" + predictor + \" not in \" + cdata.columns)\n",
    "\n",
    "    if id not in cdata.columns:\n",
    "        sys.exit(\"Error: \" + id + \" not in \" + cdata.columns)\n",
    "\n",
    "    if fid not in cdata.columns:\n",
    "        sys.exit(\"Error: \" + fid + \" not in \" + cdata.columns)\n",
    "\n",
    "    # If the pheno file exists read it, otherwise create it and save it\n",
    "    if os.path.isfile(phenofile):\n",
    "        phenoDat = pd.read_csv(phenofile, header=0, delim_whitespace=True)\n",
    "    else:\n",
    "        phenoDat = cdata[[fid, id, predictor]]\n",
    "\n",
    "        # write it\n",
    "        phenoDat.to_csv(phenofile, index=False, sep=' ')\n",
    "\n",
    "    cdata = cdata[[c for c in cdata.columns if c != predictor]]\n",
    "\n",
    "    # Create the handler\n",
    "    genoHandler = {}\n",
    "    genoHandler[\"geno\"] = geno\n",
    "    genoHandler[\"pheno\"] = pheno\n",
    "    genoHandler[\"covs\"] = covs\n",
    "    genoHandler[\"id\"] = id\n",
    "    genoHandler[\"fid\"] = fid\n",
    "    genoHandler[\"covsDat\"] = cdata\n",
    "    genoHandler[\"phenoDat\"] = phenoDat\n",
    "    genoHandler[\"Class\"] = predictor\n",
    "\n",
    "    return(genoHandler)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## GetPartitionsFromHandler function\n",
    "Function that split our data depending on the how parameter.\n",
    "\n",
    "Parameters:\n",
    "\n",
    "&emsp;__genoHandler__: A handler obtained by calling to `getHandlerToGenotypeData()`.<br>\n",
    "&emsp;__path2plink__: Path to plink<br>\n",
    "&emsp;__workPath__: String with your workpath.<br>\n",
    "&emsp;__how__: We can get a sample with \"holdout\" and \"k-fold cv\"<br>\n",
    "&emsp;__k__: The number of folds when we use cross-validation<br>\n",
    "&emsp;**p**: The proportion of data that does to training<br>\n",
    "\n",
    "Return value:\n",
    "\n",
    "&emsp;The same handler, with two new fields, `trainFolds` and `testFolds` with the indexes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "def getPartitionsFromHandler(genoHandler,\n",
    "                             path2plink,\n",
    "                             workPath=None,\n",
    "                             how=\"k-fold cv\",\n",
    "                             k=10,\n",
    "                             p=0.75):\n",
    "\n",
    "    # Verify if the value of how is a valid value\n",
    "    assert how == \"k-fold cv\" or how == \"holdout\"\n",
    "\n",
    "    genoHandler[\"plan\"] = how\n",
    "    pred = genoHandler[\"Class\"]  # Accessing the column given by the user\n",
    "\n",
    "    if how == \"holdout\":\n",
    "        # Split the data and save it\n",
    "        train, test = train_test_split(genoHandler[\"phenoDat\"][pred], test_size=(1 - p))\n",
    "        hoFold = {\"train1\": train, \"test1\": test}\n",
    "        genoHandler[\"folds\"] = hoFold\n",
    "        genoHandler[\"nfolds\"] = 1\n",
    "        genoHandler[\"trainFolds\"] = {}\n",
    "        genoHandler[\"testFolds\"] = {}\n",
    "        genoHandler[\"trainFolds\"][\"train1\"] = train\n",
    "        genoHandler[\"testFolds\"][\"test1\"] = test\n",
    "    else:\n",
    "        # Do cross-validation in data and save it\n",
    "        folds = KFold(n_splits=k)\n",
    "        genoHandler[\"folds\"] = folds\n",
    "        genoHandler[\"nfolds\"] = k\n",
    "\n",
    "        trainFolds = {}  # list with the indexes used for train in each one of the k cross-validations\n",
    "        testFolds = {}  # list with the indexes saved for test in each one of the k cross-validations\n",
    "\n",
    "        i = 1\n",
    "        for train_index, test_index in folds.split(genoHandler[\"phenoDat\"][pred]):\n",
    "            trainFolds[\"train\" + i] = train_index\n",
    "            testFolds[\"test\" + i] = test_index\n",
    "            i = i + 1\n",
    "\n",
    "        genoHandler[\"trainFolds\"] = trainFolds\n",
    "        genoHandler[\"testFolds\"] = testFolds\n",
    "\n",
    "    genoHandler = genDataFileNames(genoHandler=genoHandler, path2plink=path2plink)\n",
    "\n",
    "    return(genoHandler)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## GenDataFileNames function\n",
    "Function that generate the fields `trainFolds` and `testFolds` with the indexes in the handler.\n",
    "\n",
    "Parameters:\n",
    "\n",
    "&emsp;__genoHandler__: A handler obtained by calling to `getHandlerToGenotypeData()`.<br>\n",
    "&emsp;__onlyFold__: -1 if there are more than 1 fold, otherwise indicate the number of the fold (By default -1).<br>\n",
    "&emsp;__path2plink__: Path to plink.<br>\n",
    "&emsp;__which_to_create__: Data file names you want to generate (By default [\"train\", \"test\"])<br>\n",
    "\n",
    "Return value:\n",
    "\n",
    "&emsp;The same handler, with two new fields, `trainFolds` and `testFolds` with the indexes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "def genDataFileNames(genoHandler,\n",
    "                     onlyFold=-1,\n",
    "                     path2plink=\"~/genoml-core-master/otherPackages/\",\n",
    "                     which_to_create=[\"train\", \"test\"]):\n",
    "\n",
    "    workPath = os.path.dirname(genoHandler[\"pheno\"])\n",
    "\n",
    "    # Select the indexes\n",
    "    if onlyFold > 0:\n",
    "        indexes = onlyFold\n",
    "    else:\n",
    "        indexes = range(1, genoHandler[\"nfolds\"] + 1)\n",
    "\n",
    "    for tt in which_to_create:\n",
    "        key = tt + \"FoldFiles\"\n",
    "        genoHandler[key] = {}\n",
    "\n",
    "    for i in indexes:\n",
    "\n",
    "        for tt in which_to_create:\n",
    "            foldFiles = {}\n",
    "            key = tt + \"FoldFiles\"\n",
    "            dirFold = workPath + \"/\" + \"Fold\" + tt + str(i) + \"/\"\n",
    "\n",
    "            try:\n",
    "                os.mkdir(dirFold)\n",
    "            except Exception:\n",
    "                pass\n",
    "\n",
    "            splitIdx = tt + str(i)\n",
    "\n",
    "            # dataframe with the covariates of the individuals given by this fold\n",
    "            covsFoldDT = genoHandler[\"covsDat\"].iloc[genoHandler[tt + \"Folds\"][splitIdx].index, ]\n",
    "            covsFold = \"COVS_\" + tt + str(i)\n",
    "            covsFile = dirFold + \"/\" + covsFold + \".cov\"\n",
    "\n",
    "            # write it\n",
    "            covsFoldDT.to_csv(covsFile, index=False, sep=' ')\n",
    "\n",
    "            foldFiles[\"covsFile\"] = covsFile\n",
    "\n",
    "            genoFold = os.path.basename(genoHandler[\"geno\"]) + tt + \".\" + str(i)\n",
    "            phenoFold = os.path.basename(genoHandler[\"pheno\"]) + str(i)\n",
    "\n",
    "            # generation of the ids file\n",
    "            idsDT = genoHandler[\"covsDat\"].iloc[genoHandler[tt + \"Folds\"][splitIdx].index, ][[genoHandler[\"fid\"], genoHandler[\"id\"]]]\n",
    "            idsfile = dirFold + \"/\" + os.path.basename(genoHandler[\"covs\"]) + tt + \".ids\"\n",
    "\n",
    "            # write it\n",
    "            idsDT.to_csv(idsfile, index=False, sep=' ')\n",
    "\n",
    "            foldFiles[\"idsFile\"] = idsfile\n",
    "\n",
    "            # Generating the phenotype files\n",
    "            phenoFoldDT = genoHandler[\"phenoDat\"].iloc[genoHandler[tt + \"Folds\"][splitIdx].index, ]\n",
    "            phenofile = dirFold + \"/\" + phenoFold + tt + \".pheno\"\n",
    "\n",
    "            # write it\n",
    "            phenoFoldDT.to_csv(phenofile, index=False, sep=' ')\n",
    "\n",
    "            foldFiles[\"phenoFile\"] = phenofile\n",
    "\n",
    "            # Generating the genotype files\n",
    "            genofile = dirFold + \"/\" + genoFold\n",
    "            command = path2plink + \"plink --bfile \" + genoHandler[\"geno\"] + \" --keep \" + idsfile + \" --make-bed --out \" + genofile\n",
    "            foldFiles[\"genoFile\"] = genofile\n",
    "            foldFiles[\"genoFileCommand\"] = command\n",
    "            foldFiles[\"filesCreated\"] = False\n",
    "            genoHandler[key][i] = foldFiles\n",
    "\n",
    "    return(genoHandler)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## GenDataFromHandler function\n",
    "Function that create the geno files for each fold.\n",
    "\n",
    "Parameters:\n",
    "\n",
    "&emsp;__genoHandler__: A handler obtained by calling to `getHandlerToGenotypeData()`.<br>\n",
    "&emsp;__onlyFold__: -1 if there are more than 1 fold, otherwise indicate the number of the fold (By default -1).<br>\n",
    "&emsp;__which_to_create__: Data file names you want to generate (By default [\"train\", \"test\"])<br>\n",
    "&emsp;__lazy__: If True, does't create the files if they are already created (By default False).<br>\n",
    "\n",
    "Return value:\n",
    "\n",
    "&emsp;The same handler, with the field `filesCreated` = True."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "def genDataFromHandler(genoHandler,\n",
    "                       onlyFold=-1,\n",
    "                       which_to_create=[\"train\", \"test\"],\n",
    "                       lazy=False):\n",
    "\n",
    "    # Generate father's files first\n",
    "    if \"father\" in genoHandler.keys() and genoHandler[\"father\"] is not None:\n",
    "        genDataFromHandler(genoHandler[\"father\"], onlyFold, which_to_create, lazy)\n",
    "\n",
    "    if \"nfolds\" in genoHandler.keys() and genoHandler[\"nfolds\"] is not None:\n",
    "        # Initialize indexes\n",
    "        if onlyFold > 0:\n",
    "            indexes = onlyFold\n",
    "        else:\n",
    "            indexes = range(1, genoHandler[\"nfolds\"] + 1)\n",
    "        \n",
    "        for i in indexes:\n",
    "            for tt in which_to_create:\n",
    "                key = tt + \"FoldFiles\"\n",
    "\n",
    "                if key in genoHandler.keys() and i in genoHandler[key].keys() and genoHandler[key][i] is not None:\n",
    "                    if lazy: # Skip if files are already created\n",
    "                        if not os.path.isfile(genoHandler[key][i][\"genoFile\"] + \".bim\"):\n",
    "                            print(\"Creating genotype data for fold \" + str(i) + \" and \" + tt + \" with command \" + genoHandler[key][i][\"genoFileCommand\"] + \"\\n\")\n",
    "                            os.system(genoHandler[key][i][\"genoFileCommand\"])\n",
    "                            genoHandler[key][i][\"filesCreated\"] = True\n",
    "                        else:\n",
    "                            print(genoHandler[key][i][\"genoFile\"] + \" already created, skipping\\n\")\n",
    "                    else: # Always create files\n",
    "                        if genoHandler[key][i][\"genoFileCommand\"] is not None:\n",
    "                            print(\"Creating genotype data for fold \" + str(i) + \" and \" + tt + \" with command \" + genoHandler[key][i][\"genoFileCommand\"] + \"\\n\")\n",
    "                            os.system(genoHandler[key][i][\"genoFileCommand\"])\n",
    "                            genoHandler[key][i][\"filesCreated\"] = True\n",
    "\n",
    "    return(genoHandler)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## MostRelevantSNPs function\n",
    "Function that select from all SNPs the most relevant ones, applying prune, sblup or prsice.\n",
    "\n",
    "Parameters:\n",
    "\n",
    "&emsp;__handler__: A handler obtained by calling to `genDataFromHandler()`.<br>\n",
    "...\n",
    "\n",
    "Return value:\n",
    "\n",
    "&emsp;The same handler, with the field `snpsToPull`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "def mostRelevantSNPs(handler,\n",
    "                     path2GWAS,\n",
    "                     path2plink,\n",
    "                     reduce=\"PRSICE\",\n",
    "                     phenoScale=\"DISC\",\n",
    "                     gwas=\"RISK_noSpain.tab\",\n",
    "                     SNPcolumnatGWAS=\"SNP\",\n",
    "                     herit=None,\n",
    "                     clumpField=\"p\",\n",
    "                     addit=\"NA\",\n",
    "                     path2gcta64=None,\n",
    "                     cores=1,\n",
    "                     prune_windowSize=10000,\n",
    "                     prune_stepSize=1,\n",
    "                     prune_r2=0.1,\n",
    "                     path2PRSice=None,\n",
    "                     PRSiceexe=\"PRSice_linux\",\n",
    "                     workPath=None,\n",
    "                     force=False):\n",
    "\n",
    "    if path2gcta64 is None:\n",
    "        path2gcta64 = path2plink\n",
    "    if path2PRSice is None:\n",
    "        path2PRSice = path2plink\n",
    "\n",
    "    verifyHandler(handler)\n",
    "\n",
    "    if(path2plink != \"\"):\n",
    "        assert os.path.isdir(path2plink)\n",
    "\n",
    "    handler = genDataFromHandler(handler, lazy=True)\n",
    "    geno = os.path.basename(handler[\"geno\"])\n",
    "    pheno = os.path.basename(handler[\"pheno\"])\n",
    "\n",
    "    if workPath is None:\n",
    "        workPath = os.path.dirname(handler[\"pheno\"]) + \"/\"\n",
    "\n",
    "    path2Genotype = os.path.dirname(handler[\"geno\"]) + \"/\"\n",
    "    cov = os.path.basename(handler[\"covs\"])\n",
    "\n",
    "    # Options passed from list on draftCommandOptions.txt\n",
    "    prefix = \"g-\" + geno + \"-p-\" + pheno + \"-c-\" + cov + \"-a-\" + addit\n",
    "    fprefix = workPath + \"/\" + prefix  # CAMBIAR\n",
    "\n",
    "    if reduce == \"PRUNE\" or reduce == \"DEFAULT\":\n",
    "        command = path2plink + \"plink --bfile \" + path2Genotype + \"/\" + geno + \" --indep-pairwise \" +\\\n",
    "            prune_windowSize + \" \" + prune_stepSize + \" \" + prune_r2 + \" --out \" + fprefix +\\\n",
    "            \".temp\"\n",
    "        os.system(command)\n",
    "\n",
    "        command = path2plink + \"plink --bfile \" + path2Genotype + \"/\" + geno + \" --extract \" +\\\n",
    "            fprefix + \".temp.prune.in --recode A --out \" + fprefix + \".reduced_genos\"\n",
    "        os.system(command)\n",
    "\n",
    "        command = \"cut -f 1 \" + fprefix + \".temp.prune.in > \" + fprefix + \".reduced_genos_snpList\"\n",
    "        os.system(command)\n",
    "\n",
    "        handler[\"rgenosSnpList\"] = fprefix + \".reduced_genos_snpList\"\n",
    "        handler[\"snpsToPull\"] = handler[\"rgenosSnpList\"]\n",
    "\n",
    "        return(handler)\n",
    "\n",
    "    elif reduce == \"SBLUP\" and gwas != \"NA\" and herit is not None:\n",
    "\n",
    "        # if reduce = SBLUP, gwas is not NA and herit is not NA\n",
    "        command = \"wc -l < \" + path2Genotype + \"/\" + geno + \".bim | awk '{print $1}'\"\n",
    "        print(\"The command to run: \" + command)\n",
    "        nsnps = int(os.system(command))\n",
    "        print(\"The number of starting SNPs is \" + nsnps)\n",
    "        sbluplambda = nsnps * (1 / herit) - 1\n",
    "        print(\"We will use a sblup-lambda parameter of\" + sbluplambda)\n",
    "\n",
    "        if force or \\\n",
    "           not os.path.isfile(workPath + \"/\" + geno + \".forSblup.bim\") or \\\n",
    "           not os.path.isfile(workPath + \"/\" + geno + \".forSblup.fam\") or \\\n",
    "           not os.path.isfile(workPath + \"/\" + geno + \".forSblup.bed\"):\n",
    "\n",
    "            command = path2plink + \"plink --bfile \" + path2Genotype + \"/\" + geno + \" --pheno \" +\\\n",
    "                handler[\"pheno\"] + \".pheno --make-bed --out \" + workPath + \"/\" + geno + \".forSblup\"\n",
    "            os.system(command)\n",
    "\n",
    "        # Note that the GWAS for cojo analisys must be in the form\n",
    "        # SNP A1 A2 freq b se p N\n",
    "        # rs1001 A G 0.8493 0.0024 0.0055 0.6653 129850\n",
    "        # rs1002 C G 0.0306 0.0034 0.0115 0.7659 129799\n",
    "        # rs1003 A C 0.5128 0.0045 0.0038 0.2319 129830\n",
    "        command = path2gcta64 + \"/gcta64 --bfile \" + workPath + \"/\" + geno + \".forSblup --cojo-file \" +\\\n",
    "            path2GWAS + \"/\" + gwas + \" --cojo-sblup \" + sbluplambda + \" --cojo-wind 10000 --maf 0.01 --chr 22 --thread-num \" +\\\n",
    "            cores + \" --out \" + fprefix + \".temp\"\n",
    "        os.system(command)\n",
    "\n",
    "        # Load SBLUP results\n",
    "        sblupdata = pd.read_csv(fprefix + \".temp.sblup.cojo\", header=None, delim_whitespace=True)\n",
    "        # Start filters for sign matching and abs > 1 in sblup estimates to get ~25% data\n",
    "        sblupdata[\"match\"] = 1 if np.sign(sblupdata[\"V3\"]) == np.sign(sblupdata[\"V4\"]) else 0\n",
    "        sblupdata = sblupdata[sblupdata[\"match\"] == 1 and abs(sblupdata[\"V4\"]) > 1]\n",
    "        sblupdata.columns = [\"SNP\", \"effectAllele\", \"gwasBeta\", \"sblupBeta\", \"effectMatch\"]\n",
    "        # Export list of SNPs to pull\n",
    "\n",
    "        sblupdata.to_csv(fprefix + \".sblupToPull\", index=False, sep='\\t')\n",
    "\n",
    "        command = path2plink + \"plink --bfile \" + workPath + \"/\" + geno + \".forSblup --extract \" +\\\n",
    "            workPath + geno + \".sblupToPull --indep-pairwise 10000 1 0.1 --out \" + fprefix + \".pruning\"\n",
    "        os.system(command)\n",
    "\n",
    "        command = path2plink + \"plink --bfile \" + workPath + geno + \".forSblup --extract \" + fprefix +\\\n",
    "            \".pruning.prune.in --recode A --out \" + fprefix + \".reduced_genos\"\n",
    "        os.system(command)\n",
    "\n",
    "        # Exports SNP list for extraction in validataion set\n",
    "        command = \"cut -f 1 \" + fprefix + \".pruning.prune.in > \" + fprefix + \".reduced_genos_snpList; rm \" +\\\n",
    "            workPath + \"/\" + geno + \".forSblup.*\"\n",
    "        os.system(command)\n",
    "        handler[\"rgenosSnpList\"] = fprefix + \".reduced_genos_snpList\"\n",
    "        handler[\"snpsToPull\"] = handler[\"rgenosSnpList\"]\n",
    "\n",
    "        return(handler)\n",
    "\n",
    "    elif reduce == \"PRSICE\" and gwas != \"NA\":\n",
    "        # if reduce = PRSICE, phenoScale is DISC, gwas is not NA, $cov = NA\n",
    "\n",
    "        covstr = \" \" if cov == \"NA\" else \" --cov-file \" + path2Genotype + \"/\" + cov + \".cov \"\n",
    "        # Second check on covariate file\n",
    "        print(\"The covstr is\" + covstr)\n",
    "\n",
    "        if covstr != \" \":\n",
    "            covs = pd.read_csv(path2Genotype + \"/\" + cov + \".cov\", header=0, delim_whitespace=True)\n",
    "\n",
    "            print(covs.shape)\n",
    "\n",
    "            if len(covs.columns) == 2:\n",
    "                covstr = \" \"\n",
    "\n",
    "        print(\"The covstr is\" + covstr)\n",
    "\n",
    "        binaryTarget = True if phenoScale == \"DISC\" else False\n",
    "\n",
    "        command = genPRSiceCommand(geno,\n",
    "                                   pheno,\n",
    "                                   covstr,\n",
    "                                   path2PRSice,\n",
    "                                   PRSiceexe,\n",
    "                                   cores,\n",
    "                                   fprefix,\n",
    "                                   path2Genotype,\n",
    "                                   path2GWAS,\n",
    "                                   gwas,\n",
    "                                   binaryTarget,\n",
    "                                   gwasDef=\" --beta --snp MarkerName --A1 Allele1 --A2 Allele2 --stat Effect --se StdErr --pvalue P-value\")\n",
    "\n",
    "        print(\"The command to run: \" + command)\n",
    "        os.system(command)\n",
    "\n",
    "        command = \"cut -f 2 \" + fprefix + \".temp.snp > \" + fprefix + \".temp.snpsToPull\"\n",
    "        print(\"The command to run: \" + command)\n",
    "        os.system(command)\n",
    "\n",
    "        command = \"awk 'NR == 2 {print $3}' \" + fprefix + \".temp.summary\"\n",
    "        print(\"The command to run: \" + command)\n",
    "        try:\n",
    "            thresh = float(subprocess.check_output([\"awk\", 'NR == 2 {print $3}', fprefix + \".temp.summary\"]))\n",
    "        except subprocess.CalledProcessError as e:\n",
    "            sys.exit(e.output)\n",
    "\n",
    "        command = clumpCommand(geno, gwas, thresh, fprefix, path2plink, path2Genotype, path2GWAS, SNPcolumnatGWAS, clumpField)\n",
    "        print(\"The command to run: \" + command)\n",
    "        os.system(command)\n",
    "\n",
    "        command = \"cut -f 3 \" + fprefix + \".tempClumps.clumped > \" + fprefix + \".temp.snpsToPull2\"\n",
    "        print(\"The command to run: \" + command)\n",
    "        os.system(command)\n",
    "\n",
    "        command = path2plink + \"plink --bfile \" + path2Genotype + geno + \" --extract \" +\\\n",
    "            fprefix + \".temp.snpsToPull2 --recode A --out \" + fprefix + \".reduced_genos\"\n",
    "\n",
    "        print(\"The COMMAND: \" + command)\n",
    "        os.system(command)\n",
    "\n",
    "        # Exports SNP list for extraction in validation set\n",
    "        command = \"cut -f 1 \" + fprefix + \".temp.snpsToPull2 > \" + fprefix + \".reduced_genos_snpList\"\n",
    "        print(\"The command to run: \" + command)\n",
    "        os.system(command)\n",
    "\n",
    "        handler[\"rgenosSnpList\"] = fprefix + \".reduced_genos_snpList\"\n",
    "        handler[\"snpsToPull\"] = fprefix + \".temp.snpsToPull2\"\n",
    "        handler[\"snpsClumped\"] = fprefix + \".tempClumps.clumped\"\n",
    "\n",
    "        return(handler)\n",
    "\n",
    "    else:\n",
    "        sys.exit(\"The combination of parameters is not right\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## GetHandlerFromFold function\n",
    "Function that creates the handler from the fold.\n",
    "\n",
    "Parameters:\n",
    "\n",
    "&emsp;__handler__: A handler obtained by calling to `genDataFromHandler()`.<br>\n",
    "...\n",
    "\n",
    "Return value:\n",
    "\n",
    "&emsp;The same handler, ready for ML."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "def getHandlerFromFold(handler, type=\"train\", index=1):\n",
    "\n",
    "    key = type + \"FoldFiles\"\n",
    "    if type == \"test\":\n",
    "        assert handler[\"testFolds\"] is not None\n",
    "        assert len(handler[\"testFolds\"]) >= index\n",
    "\n",
    "    if type == \"train\":\n",
    "        assert handler[\"trainFolds\"] is not None\n",
    "        assert len(handler[\"testFolds\"]) >= index\n",
    "\n",
    "    assert handler[key] is not None\n",
    "    \n",
    "    files = handler[key][index]\n",
    "    father = handler\n",
    "    handler = getHandlerToGenotypeData(geno=files[\"genoFile\"],\n",
    "                                       covs=re.sub(\".cov\", \"\", files[\"covsFile\"]),\n",
    "                                       predictor=handler[\"Class\"],\n",
    "                                       id=handler[\"id\"],\n",
    "                                       fid=handler[\"fid\"],\n",
    "                                       pheno=re.sub(\".pheno\", \"\", files[\"phenoFile\"]),\n",
    "                                       verify=False)\n",
    "    handler[\"father\"] = father\n",
    "    handler[\"fatherkey\"] = key\n",
    "    handler[\"fatherindex\"] = index\n",
    "\n",
    "    return(handler)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## FromSNPs2MLdata function\n",
    "Function that creates the handler ready for ML.\n",
    "\n",
    "Parameters:\n",
    "\n",
    "&emsp;__handler__: A handler obtained by calling to `genDataFromHandler()`.<br>\n",
    "...\n",
    "\n",
    "Return value:\n",
    "\n",
    "&emsp;The same handler, ready for ML."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "def fromSNPs2MLdata(handler,\n",
    "                    addit,\n",
    "                    path2plink,\n",
    "                    predictor,\n",
    "                    fsHandler=None):\n",
    "\n",
    "    # We must have done feature selection with handleSNPs before\n",
    "\n",
    "    if \"snpsToPull\" not in handler.keys() or handler[\"snpsToPull\"] is None:\n",
    "        handler[\"snpsToPull\"] = \"void\"\n",
    "    if fsHandler is None:\n",
    "        # Then we have to pull SNPs as default\n",
    "        fsHandler = handler\n",
    "\n",
    "    if \"nfolds\" in handler.keys() and handler[\"nfolds\"] is not None:\n",
    "        modes = [\"train\", \"test\"]\n",
    "    else:\n",
    "        modes = [\"train\"]\n",
    "        handler[\"nfolds\"] = 1\n",
    "\n",
    "    print(\"Number of folds here \" + str(handler[\"nfolds\"]) + \"\\n\")\n",
    "\n",
    "    handler = genDataFromHandler(handler, lazy=True)\n",
    "\n",
    "    for fold in range(1, handler[\"nfolds\"] + 1):\n",
    "        for mode in modes:\n",
    "            if mode == \"train\" and len(modes) == 1:\n",
    "                geno = os.path.basename(handler[\"geno\"])\n",
    "                pheno = os.path.basename(handler[\"pheno\"])\n",
    "                workPath = os.path.dirname(handler[\"pheno\"]) + \"/\"\n",
    "                path2Genotype = os.path.dirname(handler[\"geno\"]) + \"/\"\n",
    "                cov = os.path.basename(handler[\"covs\"])\n",
    "                prefix = \"g-\" + geno + \"-p-\" + pheno + \"-c-\" + cov + \"-a-\" + addit\n",
    "            elif mode == \"train\" and len(modes) > 1:\n",
    "                geno = os.path.basename(handler[\"trainFoldFiles\"][fold][\"genoFile\"])\n",
    "                pheno = os.path.basename(re.sub(\".pheno\", \"\", handler[\"trainFoldFiles\"][fold][\"phenoFile\"]))\n",
    "                workPath = os.path.dirname(handler[\"trainFoldFiles\"][fold][\"genoFile\"]) + \"/\"\n",
    "                # path2Genotype = workPath\n",
    "                cov = os.path.basename(re.sub(\".cov\", \"\", handler[\"trainFoldFiles\"][fold][\"covsFile\"]))\n",
    "                prefix = \"g-\" + geno + \"-p-\" + pheno + \"-c-\" + cov + \"-a-\" + addit\n",
    "            elif mode == \"test\" and len(modes) > 1:\n",
    "                geno = os.path.basename(handler[\"testFoldFiles\"][fold][\"genoFile\"])\n",
    "                pheno = os.path.basename(re.sub(\".pheno\", \"\", handler[\"testFoldFiles\"][fold][\"phenoFile\"]))\n",
    "                workPath = os.path.dirname(handler[\"testFoldFiles\"][fold][\"genoFile\"]) + \"/\"\n",
    "                # path2Genotype = workPath\n",
    "                cov = os.path.basename(re.sub(\".cov\", \"\", handler[\"testFoldFiles\"][fold][\"covsFile\"]))\n",
    "\n",
    "                if handler[\"snpsToPull\"] == fsHandler[\"snpsToPull\"]:\n",
    "                    # We need the previous prefix\n",
    "                    genotrn = os.path.basename(handler[\"trainFoldFiles\"][fold][\"genoFile\"])\n",
    "                    phenotrn = os.path.basename(re.sub(\".pheno\", \"\", handler[\"trainFoldFiles\"][fold][\"phenoFile\"]))\n",
    "                    covtrn = os.path.basename(re.sub(\".cov\", \"\", handler[\"trainFoldFiles\"][fold][\"covsFile\"]))\n",
    "                    workPathtrn = os.path.dirname(handler[\"trainFoldFiles\"][fold][\"genoFile\"]) + \"/\"\n",
    "                    previous = \"g-\" + genotrn + \"-p-\" + phenotrn + \"-c-\" + covtrn + \"-a-\" + addit\n",
    "                    prefix = \"g-\" + geno + \"-p-\" + pheno + \"-c-\" + cov + \"-a-\" + addit\n",
    "\n",
    "                    command = path2plink + \"plink --bfile \" + workPath + \"/\" + geno + \" --keep \" +\\\n",
    "                        workPath + \"/\" + cov + \".cov\" + \" --extract \" + workPathtrn + \"/\" + previous +\\\n",
    "                        \".reduced_genos_snpList --recode A --out \" + workPath + \"/\" + prefix + \".reduced_genos\"\n",
    "\n",
    "                    print(\"Running command \" + command + \"\\n\")\n",
    "                    os.system(command)\n",
    "\n",
    "            else:\n",
    "                sys.exit()\n",
    "\n",
    "            print(handler[\"snpsToPull\"])\n",
    "            print(fsHandler[\"snpsToPull\"])\n",
    "\n",
    "            if handler[\"snpsToPull\"] != fsHandler[\"snpsToPull\"]:\n",
    "                print(\"We are going to generate a SNP list from a SNP pool selected outside this handler\")\n",
    "                command = path2plink + \"plink --bfile \" + workPath + geno + \" --extract \" +\\\n",
    "                    xstr(fsHandler[\"snpsToPull\"]) + \" --recode A --out \" + workPath + prefix + \".reduced_genos\"\n",
    "                os.system(command)\n",
    "                command = \"cut -f 1 \" + xstr(fsHandler[\"snpsToPull\"]) + \" > \" + workPath + prefix + \".reduced_genos_snpList\"\n",
    "                os.system(command)\n",
    "                handler[\"snpsToPull\"] = fsHandler[\"snpsToPull\"]\n",
    "\n",
    "            # now decide what to merge\n",
    "            genoPheno = 2\n",
    "            addCov = 0 if cov == \"NA\" else 1\n",
    "            addAddit = 0 if addit == \"NA\" else 1\n",
    "            nFiles = genoPheno + addCov + addAddit  # this specifies the number of files to merge\n",
    "            print(\"MERGING \" + str(nFiles) + \" FILES\")\n",
    "            genotypeInput = workPath + prefix + \".reduced_genos.raw\"\n",
    "            phenoInput = workPath + pheno + \".pheno\"\n",
    "            covInput = workPath + cov + \".cov\"\n",
    "            additInput = workPath + addit + \".addit\"\n",
    "\n",
    "            genosRaw = pd.read_csv(genotypeInput, header=0, delim_whitespace=True)\n",
    "            phenoRaw = pd.read_csv(phenoInput, header=0, delim_whitespace=True)\n",
    "\n",
    "            genosRaw[\"ID\"] = genosRaw[\"FID\"] + \"_\" + genosRaw[\"IID\"]\n",
    "            phenoRaw[\"ID\"] = phenoRaw[\"FID\"] + \"_\" + phenoRaw[\"IID\"]\n",
    "            fname = workPath + prefix + \".dataForML\"\n",
    "            phenoRaw.drop([\"FID\", \"IID\"], 1, inplace=True)\n",
    "            genosRaw.drop([\"FID\", \"IID\", \"MAT\", \"PAT\", \"SEX\", \"PHENOTYPE\"], 1, inplace=True)\n",
    "\n",
    "            # run for only geno and pheno data, ie nFiles = 2\n",
    "            if nFiles == 2:\n",
    "                temp = pd.merge(phenoRaw, genosRaw, on=\"ID\")\n",
    "                temp.rename(columns={temp.columns[0]: \"PHENO\"}, inplace=True)\n",
    "                print(temp.columns)\n",
    "                temp.to_csv(fname, index=False, sep='\\t')\n",
    "\n",
    "            # run for studies that have all geno, pheno, cov and addit data availible, ie nFiles = 4\n",
    "            elif nFiles == 4:\n",
    "                covRaw = pd.read_csv(covInput, header=0, delim_whitespace=True)\n",
    "                additRaw = pd.read_csv(additInput, header=0, delim_whitespace=True)\n",
    "                covRaw[\"ID\"] = covRaw[\"FID\"] + \"_\" + covRaw[\"IID\"]\n",
    "                additRaw[\"ID\"] = additRaw[\"FID\"] + \"_\" + additRaw[\"IID\"]\n",
    "\n",
    "                covRaw.drop([\"FID\", \"IID\"], 1, inplace=True)\n",
    "                additRaw.drop([\"FID\", \"IID\"], 1, inplace=True)\n",
    "\n",
    "                temp1 = pd.merge(phenoRaw, covRaw, on=\"ID\")\n",
    "                temp2 = pd.merge(temp1, additRaw, on=\"ID\")\n",
    "                temp3 = pd.merge(temp2, genosRaw, on=\"ID\")\n",
    "                temp3.rename(columns={temp3.columns[0]: \"PHENO\"}, inplace=True)\n",
    "                print(temp3.columns)\n",
    "                fname = workPath + prefix + \".dataForML\"\n",
    "                temp3.to_csv(fname, index=False, sep='\\t')\n",
    "                temp = temp3\n",
    "\n",
    "            # run for studies that have all geno, pheno and cov data availible (addit is missing), ie nFiles = 3\n",
    "            elif nFiles == 3 and addit == \"NA\":\n",
    "                otherRaw = pd.read_csv(covInput, header=0, delim_whitespace=True)\n",
    "                otherRaw[\"ID\"] = otherRaw[\"FID\"] + \"_\" + otherRaw[\"IID\"]\n",
    "\n",
    "                otherRaw.drop([\"FID\", \"IID\"], 1, inplace=True)\n",
    "                temp1 = pd.merge(phenoRaw, otherRaw, on=\"ID\")\n",
    "                temp2 = pd.merge(temp1, genosRaw, on=\"ID\")\n",
    "                temp2.rename(columns={temp2.columns[0]: \"PHENO\"}, inplace=True)\n",
    "                print(temp2.columns)\n",
    "                fname = workPath + prefix + \".dataForML\"\n",
    "                temp2.to_csv(fname, index=False, sep='\\t')\n",
    "                temp = temp2\n",
    "\n",
    "            # run for studies that have all geno, pheno and addit data availible (cov is missing), ie nFiles = 3\n",
    "            elif nFiles == 3 and cov == \"NA\":\n",
    "                otherRaw = pd.read_csv(additInput, header=0, delim_whitespace=True)\n",
    "                otherRaw[\"ID\"] = otherRaw[\"FID\"] + \"_\" + otherRaw[\"IID\"]\n",
    "\n",
    "                otherRaw.drop([\"FID\", \"IID\"], 1, inplace=True)\n",
    "\n",
    "                temp1 = pd.merge(phenoRaw, otherRaw, on=\"ID\")\n",
    "                temp2 = pd.merge(temp1, genosRaw, on=\"ID\")\n",
    "                temp2.rename(columns={temp2.columns[0]: \"PHENO\"}, inplace=True)\n",
    "                print(temp2.columns)\n",
    "                fname = workPath + prefix + \".dataForML\"\n",
    "                temp2.to_csv(fname, index=False, sep='\\t')\n",
    "                temp = temp2\n",
    "\n",
    "            handler[mode + str(fold) + \"mldata\"] = fname\n",
    "\n",
    "            print(\"First 100 variable names for your file below, the rest are likely just more genotypes...\")\n",
    "            print(temp.columns[0:100])\n",
    "            print(\"... and the last 100 variable names for your file below...\")\n",
    "            print(temp.columns[-100:0])\n",
    "            print(\"Your final file has \" + str(len(temp[\"ID\"])) + \" samples, and \" + str(len(temp.columns)) + \" predictors for analysis\")\n",
    "\n",
    "    return(handler)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## VerifyHandler function\n",
    "Auxiliary function that verifies if the handler is ok.\n",
    "\n",
    "Parameters:\n",
    "\n",
    "&emsp;__h__: The handler to verify.<br>\n",
    "&emsp;**level**: 1 if you havent split the data, 2 in other case.<br>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "def verifyHandler(h,\n",
    "                  level=1):\n",
    "    if(level >= 1):\n",
    "        assert os.path.isfile(h[\"geno\"] + \".bed\")\n",
    "        assert os.path.isfile(h[\"geno\"] + \".bim\")\n",
    "        assert os.path.isfile(h[\"geno\"] + \".fam\")\n",
    "        assert os.path.isfile(h[\"covs\"] + \".cov\")\n",
    "        cdata = pd.read_csv(h[\"covs\"] + \".cov\", header=0, delim_whitespace=True)\n",
    "        assert h[\"id\"] in cdata.columns\n",
    "        assert os.path.isfile(h[\"pheno\"] + \".pheno\")\n",
    "\n",
    "    if(level >= 2):\n",
    "        assert h[\"trainFolds\"] is not None\n",
    "        assert h[\"testFolds\"] is not None\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## GenPRSiceCommand function\n",
    "Function that creates the command which aplies PRSice to the data.\n",
    "\n",
    "Parameters:\n",
    "\n",
    "&emsp;__handler__: A handler obtained by calling to `genDataFromHandler()`.<br>\n",
    "...\n",
    "\n",
    "Return value:\n",
    "\n",
    "&emsp;The command."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "def genPRSiceCommand(geno,\n",
    "                     pheno,\n",
    "                     covstr,\n",
    "                     path2PRSice,\n",
    "                     PRSiceexe,\n",
    "                     cores,\n",
    "                     fprefix,\n",
    "                     path2Genotype,\n",
    "                     path2GWAS,\n",
    "                     gwas,\n",
    "                     binaryTarget,\n",
    "                     barLevels=\"5E-8,4E-8,3E-8,2E-8,1E-8,9E-7,8E-7,7E-7,6E-7,5E-7,4E-7,3E-7,2E-7,1E-7,9E-6,8E-6,7E-6,6E-6,5E-6,4E-6,3E-6,2E-6,1E-6,9E-5,8E-5,7E-5,6E-5,5E-5,4E-5,3E-5,2E-5,1E-5,9E-4,8E-4,7E-4,6E-4,5E-4,4E-4,3E-4,2E-4,1E-4,9E-3,8E-3,7E-3,6E-3,5E-3,4E-3,3E-3,2E-3,1E-3,9E-2,8E-2,7E-2,6E-2,5E-2,4E-2,3E-2,2E-2,1E-2,9E-1,8E-1,7E-1,6E-1,5E-1,4E-1,3E-1,2E-1,1E-1,1 \",\n",
    "                     gwasDef=\" --beta --snp SNP --A1 A1 --A2 A2 --stat b --se se --pvalue p\"):\n",
    "\n",
    "    return(\"Rscript \" + path2PRSice + \"PRSice.R --binary-target T --prsice \" + path2PRSice + PRSiceexe +\n",
    "           \" -n \" + xstr(cores) + \" --out \" + fprefix + \".temp --pheno-file \" + path2Genotype + \"/\" +\n",
    "           pheno + \".pheno -t \" + path2Genotype + \"/\" + geno + \" -b \" + path2GWAS + \"/\" + gwas + covstr +\n",
    "           \" --print-snp --score std --perm 10000 \" + \" --bar-levels \" + barLevels +\n",
    "           \" --fastscore --binary-target \" + str(binaryTarget) + gwasDef)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## ClumpCommand function\n",
    "Function that creates the command which aplies clump to the data.\n",
    "\n",
    "Parameters:\n",
    "\n",
    "&emsp;__handler__: A handler obtained by calling to `genDataFromHandler()`.<br>\n",
    "...\n",
    "\n",
    "Return value:\n",
    "\n",
    "&emsp;The command."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "def clumpCommand(geno,\n",
    "                 gwas,\n",
    "                 thresh,\n",
    "                 fprefix,\n",
    "                 path2plink,\n",
    "                 path2Genotype,\n",
    "                 path2GWAS,\n",
    "                 SNPcolumnatGWAS,\n",
    "                 clumpField):\n",
    "\n",
    "    return(path2plink + \"plink --bfile \" + path2Genotype + \"/\" + geno + \" --extract \" + fprefix +\n",
    "           \".temp.snpsToPull --clump \" + path2GWAS + \"/\" + gwas + \" --clump-p1 \" + str(thresh) + \" --clump-p2 \" +\n",
    "           str(thresh) + \" --clump-snp-field \" + str(SNPcolumnatGWAS) + \" --clump-field \" + str(clumpField) +\n",
    "           \" --clump-r2 0.1 --clump-kb 250 --out \" + str(fprefix) + \".tempClumps\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## FromGenoToMLdata function\n",
    "Function that creates the handler of the data ready to ML from the geno files.\n",
    "\n",
    "Parameters:\n",
    "\n",
    "&emsp;__workPath__: A handler obtained by calling to `genDataFromHandler()`.<br>\n",
    "...\n",
    "\n",
    "Return value:\n",
    "\n",
    "&emsp;A handler of the data for ML."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def fromGenoToMLdata(workPath,\n",
    "                     iter,\n",
    "                     path2Geno=\"/home/users/gsit/juanbot/JUAN_SpanishGWAS/UNRELATED.SPAIN4.HARDCALLS.Rsq0.8\",\n",
    "                     path2Covs=\"/home/users/gsit/juanbot/JUAN_SpanishGWAS/COVS_SPAIN\",\n",
    "                     predictor=\"DISEASE\",\n",
    "                     path2GWAS=\"/home/users/gsit/juanbot/JUAN_SpanishGWAS/toJuanNov7th2018/\",\n",
    "                     path2PRSice=\"/home/users/gsit/juanbot/genoml-core/otherPackages/\",\n",
    "                     path2plink=\"\",\n",
    "                     path2Pheno=None,\n",
    "                     snpsSpain=None):\n",
    "\n",
    "    if path2Pheno is None:\n",
    "        path2Pheno = workPath + \"/MyPhenotype\"\n",
    "\n",
    "    h = getHandlerToGenotypeData(geno=path2Geno,\n",
    "                                 covs=path2Covs,\n",
    "                                 id=\"IID\",\n",
    "                                 fid=\"FID\",\n",
    "                                 predictor=predictor,\n",
    "                                 # With this we assure everything will be written under workPath\n",
    "                                 pheno=path2Pheno)\n",
    "\n",
    "    holdout = getPartitionsFromHandler(genoHandler=h,\n",
    "                                       workPath=workPath,\n",
    "                                       path2plink=path2plink,\n",
    "                                       how=\"holdout\",\n",
    "                                       p=0.75)\n",
    "\n",
    "    holdout = genDataFromHandler(holdout, lazy=True)\n",
    "\n",
    "    # Save holdout\n",
    "    with open(workPath + 'holdout.pydat', 'wb') as holdout_file:\n",
    "        pickle.dump(holdout, holdout_file)\n",
    "\n",
    "    # Load holdout\n",
    "    with open(workPath + 'holdout.pydat', 'rb') as holdout_file:\n",
    "        holdout = pickle.load(holdout_file)\n",
    "\n",
    "    # If the current repo is the spanish, apply PRSice to reduce variables\n",
    "    if iter == 1:\n",
    "        handlerSNPs = mostRelevantSNPs(handler=getHandlerFromFold(handler=holdout, type=\"train\", index=1),\n",
    "                                       path2plink=path2plink,\n",
    "                                       gwas=\"RISK_noSpain_MAF0.05.tab\",\n",
    "                                       path2GWAS=path2GWAS,\n",
    "                                       PRSiceexe=\"PRSice_linux\",\n",
    "                                       path2PRSice=path2PRSice,\n",
    "                                       clumpField=\"P-value\",\n",
    "                                       SNPcolumnatGWAS=\"MarkerName\")\n",
    "    else:  # If not, extract only the SNPs selected from the spanish repo\n",
    "        handlerSNPs = holdout\n",
    "        addit = \"NA\"\n",
    "        geno = os.path.basename(handlerSNPs[\"geno\"])\n",
    "        pheno = os.path.basename(handlerSNPs[\"pheno\"])\n",
    "        cov = os.path.basename(handlerSNPs[\"covs\"])\n",
    "        path2Genotype = os.path.dirname(handlerSNPs[\"geno\"]) + \"/\"\n",
    "        prefix = \"g-\" + geno + \"-p-\" + pheno + \"-c-\" + cov + \"-a-\" + addit\n",
    "        fprefix = workPath + \"/\" + prefix\n",
    "        handlerSNPs[\"snpsToPull\"] = snpsSpain\n",
    "        command = path2plink + \"plink --bfile \" + path2Genotype + geno + \" --extract \" +\\\n",
    "            xstr(handlerSNPs[\"snpsToPull\"]) + \" --recode A --out \" + fprefix + \".reduced_genos\"\n",
    "        os.system(command)\n",
    "        # exports SNP list for extraction in validation set\n",
    "        command = \"cut -f 1 \" + xstr(handlerSNPs[\"snpsToPull\"]) + \" > \" + fprefix + \".reduced_genos_snpList\"\n",
    "        os.system(command)\n",
    "        handlerSNPs[\"rgenosSnpList\"] = fprefix + \".reduced_genos_snpList\"\n",
    "\n",
    "    # Save handlerSNPs\n",
    "    with open(workPath + 'handlerSNPs.pydat', 'wb') as handlerSNPs_file:\n",
    "        pickle.dump(handlerSNPs, handlerSNPs_file)\n",
    "\n",
    "    # Load handlerSNPs\n",
    "    with open(workPath + 'handlerSNPs.pydat', 'rb') as handlerSNPs_file:\n",
    "        handlerSNPs = pickle.load(handlerSNPs_file)\n",
    "\n",
    "    # generate mldata for the repository\n",
    "    mldatahandler = fromSNPs2MLdata(handler=holdout, addit=\"NA\", path2plink=path2plink, predictor=predictor, fsHandler=handlerSNPs)\n",
    "\n",
    "    # Save handlerSNPs\n",
    "    with open(workPath + \"mldatahandler.pydat\", 'wb') as mldatahandler_file:\n",
    "        pickle.dump(mldatahandler, mldatahandler_file)\n",
    "\n",
    "    # Load handlerSNPs\n",
    "    with open(workPath + 'mldatahandler.pydat', 'rb') as mldatahandler_file:\n",
    "        mldatahandler = pickle.load(mldatahandler_file)\n",
    "\n",
    "    # These are the two ML datasets generated from the genotype data and the\n",
    "    # feature selection\n",
    "    print(\"Your train dataset is at \" + mldatahandler[\"train1mldata\"])\n",
    "    print(\"Your test dataset is at\" + mldatahandler[\"test1mldata\"])\n",
    "\n",
    "    return mldatahandler\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
